Integrate profiler in perf_bench_test.go (#149)
ease-lab/vhive: docs/profiling.md

# vHive profiling tool

The tool focus on the performance characteristics of vHive framework and identify potential 
performance bottleneck of colocating VMs on a single host. The motivation of this tool is to 

> improve the parallelism of serverless architecture and commercial benefits for cloud vendors.

Reviewer: the last sentence doesn't make a lot of sense

Author: Done

> improve the parallelism of serverless architecture and commercial benefits for cloud vendors.

Reviewer 2: Please rephrase the last part. It sounds like we are employees of cloud providers.

## Methodology

The tool is for benchmarking vHive instances by recording hardware counters, requests per second
 (RPS) and execution time for functions running inside instances.

The tool includes the following components:
- A loader function issues requests to vHive step by step until the RPS reaches the maximum. At 
  each step, the function issues a percentage of the maximum RPS (e.g., 5% of the maximum RPS) 
  and profile counters. If tail latency violates 10x image unloaded service time at a step, the 

>   function returns the metric before it.

Reviewer 2: What happens after the the tail latency is violated? Please clarify this part a bit.
To whom does the function issue the percentage of maximum RPS? And what happens to them after?

Author: Done

- A single latency measurement goroutine measures the latency by injecting requests in 
  Round-Robin and measures runtimes of requests to compute the mean latency and the 
  90-percentile latency.
- The tool can bind only 1 VM to a core for profiling and other VMs to a socket
- A profiler invokes [toplev](https://github.com/andikleen/pmu-tools) to collect hardware counters.
- A plotter plots line charts. The X-axis is the number of VMs and Y-axis is the value of a metric.

> A load step is divided into three phases: warm-up, profiling and cool-down. During the profile 

Reviewer 2: what do warm up profile and cool-down period mean? what is the difference? why we need 3 modes?

Author: Done

return successfully. During the profile period, it also invokes [pmu-tools](https://github.com/andikleen/pmu-tools) 
to profile user-defined counters followed by [the TopDown method](https://ieeexplore.ieee.org/document/6844459). 
After the measurement finishes, completed RPS per core, average execution time and the average 
counters are saved in the `profile.csv`.

## Runtime Arguments
```
General:
-warmUpTime   FLOAT  The warm up time before profiling in seconds (default 5)
-profileTime  FLOAT  The profiling time in seconds (default 10)
-coolDownTime FLOAT  The cool down time after profiling in seconds (default 1)
-loadStep     FLOAT  The percentage of target RPS the benchmark loads at every step (default 5)
-funcNames    STR    Names of the functions to benchmark, separated by comma (default helloworld)
-profileCPU   INT    Bind only one VM to the CPU and profile the physical core only (default -1)
-bindSocket   INT    Bind all VMs to socket number apart from the profile CPU (default -1)
-latSamples   INT    The number of latency measurements during one profiling period (default 100)

> TestProfileSingleConfiguration:

Reviewer 2: Suggestion: TestProfileSingleConfiguration -> TestProfileFixedVMs

Author: Dmitrii thinks fixed VMs is not clear, so I changed to single configuration. The same reason for increment configuration

-vm           INT    The number of VMs (default 2)
-rps          INT    The target requests per second (default 10)

> TestProfileIncrementConfiguration:

Reviewer 2: Suggestion: TestProfileIncrementConfiguration: -> TestProfileIncrementVMs

-vmIncrStep   INT    The increment VM number (default 1)
-maxVMNum     INT    The maximum VM number (default 100)

Profiler:
-l            INT    Profile level (default 1)
-I            UINT   Print count deltas every N milliseconds (default 500)
-nodes        STR    Include or exclude nodes (with + to add, -|^ to remove,
                     comma separated list, wildcards allowed, 
                     add * to include all children/siblings, 
                     add /level to specify highest level node to match, 
                     add ^ to match related siblings and metrics, 

>                      start with ! to only include specified nodes)

Reviewer 2: Can we have more examples of the syntax? It is not easy to understand using the syntax alone.

Author: Done

# vHive profiling tool

> The tool focus on the performance characteristics of vHive framework. It uses a practical method, 

Reviewer: ```suggestion
This tool allows its users to study the performance characteristics of serverless functions
in a highly multi-tenant environment. The tool relies on 
```

> [the TopDown method](https://ieeexplore.ieee.org/document/6844459), to identify potential 
> performance bottlenecks by colocating VMs on a single host. The tool issues requests to VMs in
> Round-Robin and takes metrics from the VMs, such as RPS per CPU, hardware counters, etc. If multiple
> groups of metrics are collected, then line charts can be drawn for each attribute of metrics. An example 
> is shown in the [section](###TestProfileIncrementConfiguration-function) below.

Reviewer: ```suggestion
[the TopDown method](https://ieeexplore.ieee.org/document/6844459), to identify performance bottlenecks
that arise when colocating VMs on a single host. The tool issues requests to VMs in Round-Robin 
and collects various high-level and low level metrics, including requests-per-second (RPS)
 per core, tail latency, hardware counters, plotting the collected metrics in a set of charts.
An example chart is shown in the [section](###TestProfileIncrementConfiguration-function).
```

## Methodology

> 

Reviewer: good that you started this section with enumerating the components. however, binding is not a component is just a part of the setup. It is supplementary.

Author: Done

The tool includes the following components:

> - ***A loader function*** issues requests to vHive step by step until the RPS reaches the maximum. 
>   At each step, the function issues a percentage of the maximum RPS (e.g., 5% of the maximum RPS) 
>   and profile counters.
> - ***A latency measurement goroutine*** measures the latency by injecting requests in 
>   Round-Robin and measures runtimes of requests to compute the mean latency and the 
>   90-percentile latency.

Reviewer: ```suggestion
- ***A loader function*** issues requests to functions, running in VMs, in an open loop.
The load on VMs increases step by step until the tail latency constraint, provided by the user,
is violated. The tool collects the metrics from the last step of the load that satisfies
the latency constraint.
- ***A latency measurement goroutine*** measures the latency by injecting requests
at a low rate in  Round-Robin and measures the service time of these requests to compute 
the mean latency and the tail latency (90-percentile).
```

> - ***A bind function*** binds VMs in two ways:

Reviewer: the binding is secondary, this is just for stable and accurate measurements. You need to explain what the "profiler" is and why it is there.

Author: Done

>   -  If profile CPU ID is set, the tool allocates only one VM to the physical core where the CPU is

Reviewer: I don't understand this sentence at all. You should first say that the tool collects counters from a single core then explain why your tool binds a VM to that core, not vice versa.

Author: Done

TestProfileSingleConfiguration:
-vm           INT    The number of VMs (default 2)
-rps          INT    The target requests per second (default 10)

TestProfileIncrementConfiguration:
-vmIncrStep   INT    The increment VM number (default 1)
-maxVMNum     INT    The maximum VM number (default 100)

Profiler:
-l            INT    Profile level (default 1)
-I            UINT   Print count deltas every N milliseconds (default 500)
-nodes        STR    Include or exclude nodes (with + to add, -|^ to remove,
                     comma separated list, wildcards allowed [+Fetch_Latency,-Backend_Bound], 
                     add * to include all children/siblings [+Frontend_Bound*], 
                     add /level to specify highest level node to match [Frontend_Bound*/2], 
                     add ^ to match related siblings and metrics [^Frontend_Bound], 
                     start with ! to only include specified nodes ['!Frontend_Bound'])
```

## Pre-requisites
At the root of this repository, please run the following script to install the essential tools
for profiling and binding.
```
scripts/install_pmutool.sh 
```

## Quick-start guide
### TestProfileSingleConfiguration function

> `TestProfileSingleConfiguration` is for collecting counters from a fixed number of VMs and RPS 
> setting during the profiling period.

Reviewer: ```suggestion
`TestProfileSingleConfiguration` is for collecting the TopDown metrics from a fixed number of VMs and RPS 
setting during the profiling period.
```

Profile a single VM with `helloworld` image and 20 RPS at TopDown level 1 :
```
sudo env "PATH=$PATH" go test -v -timeout 99999s -run TestProfileSingleConfiguration \
-args -funcNames helloworld -vm 1 -rps 20 -l 1
```

> Bottleneck counters are printed out with their value at each step during 

Reviewer: ```suggestion
Bottleneck counters are printed out with their values at each step during 
```

execution, such as:
```
...
INFO[] Current RPS: 1200
INFO[] Bottleneck Backend_Bound with value 75.695000
...
```
    

> To study microarchitectural bottlenecks in more detail, profile the same configuration with sub-level counters. 
> For example, the bottleneck is in the backend bound at level 1 and now we want to profile level 2 of the backend bound:

Reviewer: ```suggestion
To study microarchitectural bottlenecks in a more detail, one needs to profile the same
configuration at the lower level. 
For example, if the bottleneck is in the backend at level 1, one should profile level 2
of the backend category:
```

```
sudo env "PATH=$PATH" go test -v -timeout 99999s -run TestProfileSingleConfiguration \ 
-args -funcNames helloworld -vm 1 -rps 20 -nodes '!+Backend_Bound*/2,+MUX'
```

### TestProfileIncrementConfiguration function

> `TestProfileIncrementConfiguration` increments the number of VMs by user-defined number until it 
> reaches the user-defined maximum. At each step, the maximum RPS is the upper bound of RPS for 
> the number of VMs at the step. For instance, there are 4 VMs with `helloworld` images. The 
> unloaded service time of the image is 1 millisecond. So, the maximum RPS is around 4000. The 
> request issue and profiling part of this function behave as same as `TestProfileSingleConfiguration`.

Reviewer: ```suggestion
`TestProfileIncrementConfiguration` increments the number of VMs by a user-defined number,
further referred to as the increment, until the number of active VMs reaches the user-defined
maximum. At each step, the maximum RPS is projected based on the independently measured
average unloaded service time of the deployed functions, and the number of cores that are
available for the VMs to run on. For instance, let us assume that there are 4 VMs running
a `helloworld` function. The unloaded service time of this function is 1 millisecond, thus
the maximum RPS is 4000, assuming that there are more than 4 cores in the CPU. 
The tool would start at 5% of 4000 RPS (i.e., 200) and collect the metrics at each step,
gradually incrementing the RPS until reaching the maximum RPS without violating
the tail latency constraint. Then, the tool starts more VMs, according to the increment's
value, and repeats this procedure.
```

Profile from 1 VMs to 32 VMs (increment step is 1) with `helloworld` image at TopDown level 1:
```
sudo env "PATH=$PATH" go test -v -timeout 99999s -run TestProfileIncrementConfiguration \ 
-args -funcNames cnn_serving -vmIncrStep 1 -maxVMNum 32 -l 1
```

> Once the profiling iteration finishes, all results are saved in the `profile.csv`. Then, the plotter retrieves 
> the contents in the file and plots according to the attributes. Here is a sample image of RPS per physical core 
> as the number of VMs increases from 1 to 32:

Reviewer: ```suggestion
Once the profiling iteration finishes, all results are saved in the `profile.csv`, and
the tool plots the TopDown metrics in a set of charts. Here is a sample image of
RPS per physical core as the number of VMs increases from 1 to 32:
```

...
```
    
To study microarchitectural bottlenecks in a more detail, one needs to profile the same
configuration at the lower level. 
For example, if the bottleneck is in the backend at level 1, one should profile level 2
of the backend category:
```
sudo env "PATH=$PATH" go test -v -timeout 99999s -run TestProfileSingleConfiguration \ 
-args -funcNames helloworld -vm 1 -rps 20 -nodes '!+Backend_Bound*/2,+MUX'
```

### TestProfileIncrementConfiguration function
`TestProfileIncrementConfiguration` increments the number of VMs by a user-defined number,
further referred to as the increment, until the number of active VMs reaches the user-defined
maximum. At each step, the maximum RPS is projected based on the independently measured
average unloaded service time of the deployed functions, and the number of cores that are
available for the VMs to run on. For instance, let us assume that there are 4 VMs running
a `helloworld` function. The unloaded service time of this function is 1 millisecond, thus
the maximum RPS is 4000, assuming that there are more than 4 cores in the CPU. 
The tool would start at 5% of 4000 RPS (i.e., 200) and collect the metrics at each step,

> gradually incrementing the RPS until reaching the maximum RPS without violating

Reviewer: ```suggestion
gradually increasing the RPS until reaching the maximum RPS without violating
```
```suggestion
gradually incrementing the RPS until reaching the maximum RPS without violating
```

framework to run. The cool-down period is because the requests are issued in Round-Robin, there are some 
requests that serve last on some VMs. So, the system runs more stably in the profiling period.

During the profile period, the loader function records the average execution time of invocations and 
how many invocations return successfully. The Profiler and latency measurement goroutine also measures 
hardware counters and latencies at this phase. 

If tail latency violates 10x image unloaded service time at an RPS step, the function stops the iteration 
and returns the metric before it. If the function reaches the maximum RPS, it returns the metric from 
the maximum RPS step. After the iteration stops, completed RPS per CPU, average execution time and the 
average counters are saved in the `profile.csv`.

For stable and accurate measurements, there are two ways of binding VMs. If all VMs are running the same image,
only one VM needs to be measured to get rid of potential noises from global measurement. User can set profile CPU 
ID and the tool allocates only one VM to the physical core of the CPU. Then, the profiler collects counters from the core.

> While the vHive is running, Other processes may interfere the framework. Therefore, all VMs can be bind to a socket. 
> If both profile CPU ID and bind socket are set, the CPU ID must be in the socket.

Reviewer: ```suggestion
  While the tool is running, one may want to bind the VMs to a single socket to exclude
  the interference of the processes that are running on the same CPU, e.g.,
  the loader functionality. If both parameters, `-profileCPUID` and `-bindSocket`,
  are defined, the former should point to a core in the same socket.
```

Author: Done

> # vHive profiling tool

Reviewer: please split the long lines (80-100 characters are MAX for a single line)

Author: Done

## Methodology

> The benchmark loads a certain amount of requests per second (RPS) to vHive for a time duration.

Reviewer: please explain the purpose(s) of this tool. What can it be used for?

then, you need to explain the setup: loader goroutines and a single latency measurement goroutine, and so on

Author: Done

> The load procedure is split into different steps from 5% to 100% of input RPS. If the tail latency at a load step violates the constraint which is 5x image service time, the tool will return the profiling result from one step before the tail latency violation.

Reviewer: 10x?

Author: Done

> A load step divides into three parts: warm-up, profiling and cool-down. During the profile period, the benchmark records the average execution time of invocations and how many invocations return successfully. It also invokes [pmu-tools](https://github.com/andikleen/pmu-tools) to monitor user-defined nodes followed by [TopDown method](https://ieeexplore.ieee.org/document/6844459). After the measurement finishes, successful RPS per core, average execution time and the average counters of nodes will save in the `profile.csv`.

Reviewer: this should come before the previous paragraph.

instead of "successful RPS", you should say completed requests per second (RPS). 

"user-defined nodes" is confusing, say "socket"

Author: Done

> 

Reviewer: you need to explain all important runtime arguments here so that a user of the tool is able to fine-tune them for their functions, server, sockets, cores, etc.

Author: Done

> ## Profile fixed number of VMs and RPS

Reviewer: You need a better section title. Maybe, "Quick-start guide". 

You should start by stating the goal of this experiment.

Author: Done

> 

Reviewer: you need to describe the pre-requisites for running this test. Particularly, say that `scripts/install_pmutool.sh` needs to be run before running any experiments.

Author: Done

Profile a single VM with `helloworld` image and 20 RPS at TopDown level 1 :
```

> sudo env "PATH=$PATH" go test -v -timeout 99999s -run TestProfileSingleConfiguration -args -funcNames helloworld -vm 1 -rps 20 -l 1

Reviewer: all these runtime arguments need to be explained in the previous section.

Author: Done

```

> bottleneck nodes will be printed out with their value. 

Reviewer: capitalize the sentence. Give a sample output.

Author: Done

> To drill down on bottlenecks, profile the same configuration with specific nodes. For example, the bottleneck is in the frontend at level 1 and now we want to profile level 2 of the frontend:

Reviewer: To drill down on bottlenecks => to study microarchitectural bottlenecks in more detail

"with specific nodes" I didn't get what you mean.

Author: Done

# vHive profiling tool

> 

Reviewer: before the methodology, you should explain what purpose the tool serves. which research goals it may help? what is the motivation for it?

Author: Done

## Methodology

The tool is for benchmarking vHive instances by recording hardware counters, requests per second
 (RPS) and execution time for functions running inside instances.

The tool includes following components: 

> - A loader function loads requests to vHive round by round (increment 5% of input RPS each round 

Reviewer: the step (call it a step not increment) is configurable, you should say it: "a step, e.g., 5% of the ...". "input RPS" is unclear

Author: Done

  by default) until input RPS and profiles performance. if tail latency violates 10x image 
  unloaded service time at a step, the function returns the metric before it.

> - A single latency measurement goroutine mimics a rare event loader that injects a request least 

Reviewer: ```suggestion
- A single latency measurement goroutine that measures the latency, by injecting a request 
```

Author: Done

>   every 500ms and measures runtimes of requests to compute mean latency and 90 percentile latency.

Reviewer: ```suggestion
  every 500ms and measures runtimes of requests to compute the mean latency and the 90-th percentile latency.
```

Author: Done

> - A bind socket function binds all VMs to socket 1 (only compatible with a 2x 16-core machine or 

Reviewer: ```suggestion
- The tool binds all VMs to socket 1 that is only compatible with a 2-socket 16-core machine,
or more sockets and more cores per socket. 
```

Author: Done

>   above).

Reviewer: ```suggestion
```

Author: Done

- A profiler invokes [toplev](https://github.com/andikleen/pmu-tools) to collect hardware counters.

> - A plotter plots line charts for a list of metrics.

Reviewer: need to be more concrete in what it plots

Author: Done

> A load step divides into three parts: warm-up, profiling and cool-down. During the profile 

Reviewer: ```suggestion
A load step is divided into three phases: warm-up, profiling and cool-down. During the profile 
```

Author: Done

period, the benchmark records the average execution time of invocations and how many invocations 
return successfully. During the profile period, it also invokes [pmu-tools](https://github.com/andikleen/pmu-tools) 
to profile user-defined counters followed by [TopDown method](https://ieeexplore.ieee.org/document/6844459). 
After the measurement finishes, completed RPS per core, average execution time and the average 

> counters will save in the `profile.csv`.

Reviewer: ```suggestion
counters are saved in the `profile.csv`.
```

Author: Done

## Runtime Arguments

> ```

Reviewer: please add the default values, I think Golang prints them if you run a test with `-h`.

Author: Done

-coolDownTime FLOAT  The cool down time after profiling in seconds
-loadStep     FLOAT  The percentage of target RPS the benchmark loads at every step
-funcNames    STR    Names of the functions to benchmark, separated by comma
-bindSocket   BOOL   Bind all VMs to socket 1 and profile one physical core only
                     (only compatible with a 2x 16-core machine or above)

TestProfileSingleConfiguration:
-vm           INT    The number of VMs
-rps          INT    The target requests per second

TestProfileIncrementConfiguration:
-vmIncrStep   INT    The increment VM number
-maxVMNum     INT    The maximum VM number

Profiler:
-l            INT    Profile level
-I            UINT   Print count deltas every N milliseconds
-nodes        STR    Include or exclude nodes (with + to add, -|^ to remove, 
                     comma separated list, wildcards allowed, 
                     add * to include all children/siblings, 
                     add /level to specify highest level node to match, 
                     add ^ to match related siblings and metrics, 
                     start with ! to only include specified nodes)
```

## Pre-requisites

> Assume you are at the root of this repository, run `scripts/install_pmutool.sh` to install

Reviewer: block code snippets are easier to copy. Please re-write

"
At the root of this repository, please run the following script to install the essential tools
for profiling and binding.
```bash
scripts/install_pmutool.sh
```
"

Author: Done

> essential tools for profiling and binding.

Reviewer: ```suggestion
```

Author: Done

## Quick-start guide

> Function `TestProfileSingleConfiguration` is for collecting counters from a fixed number of VMs and RPS setting 

Reviewer: you may want to have 2 subsections, each for TestProfileSingleConfiguration and TestProfileIncrementConfiguration

Author: Done

during profiling period.

Profile a single VM with `helloworld` image and 20 RPS at TopDown level 1 :
```
sudo env "PATH=$PATH" go test -v -timeout 99999s -run TestProfileSingleConfiguration -args -funcNames helloworld -vm 1 -rps 20 -l 1
```
Bottleneck counters will be printed out with their value at each step during 
execution, such as:
```
...
INFO[] Current RPS: 1200
INFO[] Bottleneck Backend_Bound with value 75.695000
...
```
    
To study microarchitectural bottlenecks in more detail, profile the same configuration with sub-level counters. 
For example, the bottleneck is in the backend bound at level 1 and now we want to profile level 2 of the backend bound:
```
sudo env "PATH=$PATH" go test -v -timeout 99999s -run TestProfileSingleConfiguration -args -funcNames helloworld -vm 1 -rps 20 -nodes '!+Backend_Bound*/2,+MUX'
```

> Function `TestProfileIncrementConfiguration` will load the pre-stored number of RPS and 

Reviewer: this paragraph is unclear. please assume unexperienced user and explain step-by-step.

Author: Done

increment the number of VMs step by step until it reaches the input maximum. At each 
step, this function behaves as same as `TestProfileSingleConfiguration`

Profile from 4 VMs to 32 VMs (increment step is 4) with `helloworld` image at TopDown level 1:
```
sudo env "PATH=$PATH" go test -v -timeout 99999s -run TestProfileIncrementConfiguration -args -funcNames cnn_serving -vmIncrStep 4 -maxVMNum 32 -l 1
```

> Once the profiling iteration finishes, all results will save in the `profile.csv`. Then, the plotter retrieves 

Reviewer: ```suggestion
Once the profiling iteration finishes, all results are saved in the `profile.csv`. Then, the plotter retrieves 
```

Author: Done

> the contents in the file and plots according to the attributes.

Reviewer: sample figures?

Author: Done
